packages/react/src/component.tsx

```diff
@@ -75,27 +78,36 @@ const allowedContentTextLabelGenerator = (
  *   onUploadError={(err) => console.log(err)}
  * />
  */
-export function UploadButton<TRouter extends void | FileRouter = void>(props: {
-  endpoint: EndpointHelper<TRouter>;
-  onClientUploadComplete?: (
-    res?: Awaited<ReturnType<typeof DANGEROUS__uploadFiles>>,
-  ) => void;
-  onUploadError?: (error: Error) => void;
-}) {
+export function UploadButton<
+  TRouter extends FileRouter,
+  TEndpoint extends keyof TRouter,
+>(
+  props: {
+    endpoint: TEndpoint;
+    onClientUploadComplete?: (
+      res?: Awaited<ReturnType<typeof DANGEROUS__uploadFiles>>,
+    ) => void;
+    onUploadError?: (error: Error) => void;
+  } & (inferEndpointInput<TRouter, TEndpoint> extends undefined
+    ? {}
+    : { input: inferEndpointInput<TRouter, TEndpoint> }),
+) {
```

Comment:
probable extract the type out here to a shared one that each component can extend from - duplicating it 3 times now...